src/auth/auth.service.ts

```diff
@@ -0,0 +1,132 @@
+import { userService } from "@/user/user.service";
+import { jwtTokenService } from "@/services/JwtTokenService";
+import {
+  UnauthenticatedError,
+  UnprocessableEntityError,
+} from "@/errors/http";
+import {
+  LoginForm,
+  loginFormSchema,
+  RegistrationForm,
+  registrationFormSchema,
+} from "./auth.forms";
+import { validateExistingUser, validateUserPassword } from "./auth.validators";
+import bcrypt from "bcryptjs";
+import { prisma } from "@/db/prisma";
+
+export class AuthService {
+  async register(data: RegistrationForm) {
+    const result = registrationFormSchema.safeParse(data);
+
+    if (!result.success) {
+      throw new UnprocessableEntityError({
+        errors: result.error.flatten().fieldErrors,
+      });
+    }
+
+    const { email, password, name } = data;
+
+    await validateExistingUser(email);
+
+    const hashedPassword = await bcrypt.hash(password, 10);
+    const newUser = await userService.createUser(name, email, hashedPassword);
+    const tokenPair = jwtTokenService.generateTokenPair({ userId: newUser.id });
```

Comment:
```suggestion
    const tokenPair = await jwtTokenService.generateTokenPair({ userId: newUser.id });
```

Generating token could be resource intensive operation, we probably don't want to block the event loop, especially while doing that for a pair.

Reply:
you cannot just add await and expect that function became async. 
jwt.sign has its own rules and it can be either sync or async according to their documentation and you need to write callback to make this functions async.  
We don't have a problem now with sync, so let's not complicate

Reply:
Deal!